src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -480,9 +720,361 @@ private partial class KeyValueListNode
+        private string _id;
+        private ActivityTraceId _traceId;
+        private bool _traceIdSet;
+        private ActivitySpanId _spanId;
+        private bool _spanIdSet;
+
+        // State associated with ParentId.  It can be represented by a string or by TraceId, SpanId.  
+        private string _parentId;
+        private ActivitySpanId _parentSpanId;
+        private bool _parentSpanIdSet;
+
         private bool isFinished;
         #endregion // private
     }
+
+    /// <summary>
+    /// The possibilities for the format of the ID
+    /// </summary>
+    public enum ActivityIdFormat : byte
+    {
+        Unknown,      // ID format is not known.     
+        Hierarchical, //|XXXX.XX.X_X ... see https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md#id-format
+        W3C,          // 00-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX-XXXXXXXXXXXXXXXX-XX see https://w3c.github.io/trace-context/
+    };
+
+    /// <summary>
+    /// A TraceId is the format the W3C standard requires for its ID for the entire trace. 
+    /// It represents 16 binary bytes of information, typically displayed as 32 characters
+    /// of Hexadecimal.  A TraceId is a STRUCT, and does contain the 16 bytes of binary information
+    /// so there is value in passing it by reference.   It does know how to convert to and
+    /// from its Hexadecimal string representation, tries to avoid changing formats until
+    /// it has to, and caches the string representation after it was created.   
+    /// It is mostly useful as an exchange type.  
+    /// </summary>
+    public unsafe struct ActivityTraceId : IEquatable<ActivityTraceId>
+    {
+        /// <summary>
+        /// Creates a TraceId from a 16 byte span 'idBytes'.   The bytes are copied.  
+        /// </summary>
+        public ActivityTraceId(ReadOnlySpan<byte> idData, bool isUtf8Chars = false)
+        {
+            _id1 = 0;
+            _id2 = 0;
+            _asHexString = null;
+            fixed (ulong* idPtr = &_id1)
+            {
+                var outBuff = new Span<byte>(idPtr, sizeof(ulong) * 2);
+                if (isUtf8Chars)
+                {
+                    if (idData.Length != 32)
+                        throw new ArgumentOutOfRangeException(nameof(idData));
+                    ActivityTraceId.SetSpanFromHexUtf8Chars(outBuff, idData);
+                }
+                else
+                {
+                    if (idData.Length != 16)
+                        throw new ArgumentOutOfRangeException(nameof(idData));
+                    idData.CopyTo(outBuff);
+                }
+            }
+        }
+
+        public ActivityTraceId(ReadOnlySpan<char> idData)
+        {
+            _id1 = 0;
+            _id2 = 0;
+            _asHexString = null;
+            fixed (ulong* idPtr = &_id1)
+            {
+                if (idData.Length != 32)
+                    throw new ArgumentOutOfRangeException(nameof(idData));
```

Comment:
You can move this up to the beginning of the ctor, since it's argument validation.